pkg/controller/atlasproject/atlasproject_controller.go

```diff
@@ -18,28 +18,58 @@ package atlasproject
 
 import (
 	"context"
-	"github.com/go-logr/logr"
 	mdbv1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/atlas"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/util/kube"
+	"go.uber.org/zap"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"time"
 )
 
 // AtlasProjectReconciler reconciles a AtlasProject object
 type AtlasProjectReconciler struct {
 	client.Client
-	Log    logr.Logger
+	Log    zap.SugaredLogger
 	Scheme *runtime.Scheme
 }
 
 // +kubebuilder:rbac:groups=mongodb.com.mongodb.com,resources=atlasprojects,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=mongodb.com.mongodb.com,resources=atlasprojects/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
 
 func (r *AtlasProjectReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("atlasproject", req.NamespacedName)
+	ctx := context.Background()
+	log := r.Log.With("atlasproject", req.NamespacedName)
 
-	// your logic here
+	project := &mdbv1.AtlasProject{}
+	if err := r.Client.Get(ctx, kube.ObjectKey(req.Namespace, req.Name), project); err != nil {
+		// TODO make generic (update status, log message)
+		log.Error(err, "Failed to read the AtlasProject")
+		return reconcile.Result{RequeueAfter: time.Second * 10}, nil
+	}
+
+	log.Infow("-> Starting AtlasProject reconciliation", "spec", project.Spec)
+
+	if project.Spec.ConnectionSecret == nil {
+		log.Error("So far the Connection Secret in AtlasProject is mandatory!")
+		return reconcile.Result{}, nil
+	}
+
+	connection, err := atlas.ReadConnection(r.Client, "TODO!", project.ConnectionSecretObjectKey(), log)
+	if err != nil {
+		log.Error(err, "Failed to read Atlas Connection details")
+		return reconcile.Result{RequeueAfter: time.Second * 10}, nil
+	}
+
+	if err := ensureProjectExists(connection, project, log); err != nil {
+		log.Error(err, "Failed to read the AtlasProject")
+		return reconcile.Result{RequeueAfter: time.Second * 10}, nil
```

Comment:
[nit] I usually find to define a const like `const refreshRate = 10 * time.Second` and then using that const around to be more readable/maintainable, also given how this is used multiple time in this file could make a refactor of the time easier later on

Reply:
totally agree! Planned to address this in the next PRs, this one has many TODOs about consolidating functionality. Here, for example some other things should happen: update the status of the resource, log, return - this should (and will) be handled the more generic way 👍